# Data Analysis with Python

<h3>By Priyanka Roy</h3>

<h3>Objectives: </h3>


<h4> Data Analysis:</h4>

- Using Python's different bulit-in libraries
- Read different types of files with Pandas Dataframe (.csv file, .xlsx file, etc.) <br>

<h4>Data Manipulation: </h4>

- Creating and naming the new data frame in Pandas
- Find the number of rows and columns in the dataframe
- Find the data statistics of the dataset
- Find the data types and missing values
- Rename the column names
- Remove unnecessary columns
- Analyzing the data and find answers of different kinds of questions that arises on mind while dealing with the dataset
- Arrange the dataset in a custom column order and making it more understandable and so on.

<h4> Reference(s): </h4>   <br> 
[1] <a href="https://stats.espncricinfo.com/ci/content/records/93276.html"> Actual Dataset Source </a> <br>

[2] <a href="https://github.com/priyan-2020/Test_Cricket_Analysis/blob/main/wickets.csv"> wickets.csv file </a>

<h4>Import required libraries </h4>

In [1]:
import pandas as pd
import numpy as np

<h4>Reading the wickets.csv file and showing the first 10 rows of the Dataframe </h4>

In [2]:
df= pd.read_csv("wickets.csv")

df.head(10)

Player       Span   Mat  Inns  Balls   Runs  Wkts    BBI  \
0  M Muralitharan (ICC/SL)  1992-2010   133   230  44039  18180   800   9/51   
1           SK Warne (AUS)  1992-2007   145   273  40705  17995   708   8/71   
2        JM Anderson (ENG)  2003-2021  164*   304  35079  16575   623   7/42   
3         A Kumble (INDIA)  1990-2008   132   236  40850  18355   619  10/74   
4         GD McGrath (AUS)  1993-2007   124   243  29248  12186   563   8/24   
5          SCJ Broad (ENG)  2007-2021   149   274  29863  14590   524   8/15   
6            CA Walsh (WI)  1984-2001   132   242  30019  12688   519   7/37   
7            DW Steyn (SA)  2004-2019    93   171  18608  10077   439   7/51   
8      N Kapil Dev (INDIA)  1978-1994   131   227  27740  12867   434   9/83   
9        HMRKB Herath (SL)  1999-2018    93   170  25993  12157   433  9/127   

      BBM    Ave  Econ    SR   5  10  
0  16/220  22.72  2.47  55.0  67  22  
1  12/128  25.41  2.65  57.4  37  10  
2   11/71  26.60  2.83  56.3  30   3  
3  14/149  29.65  2.69  65.9  35   8  
4   10/27  21.64  2.49  51.9  29   3  
5  11/121  27.84  2.93  56.9  18   3  
6   13/55  24.44  2.53  57.8  22   3  
7   11/60  22.95  3.24  42.3  26   5  
8  11/146  29.64  2.78  63.9  23   2  
9  14/184  28.07  2.80  60.0  34   9

<h4> Findings: </h4> <br>
Above this cell, the first column is the index no. and after that there are 14 columns altogether in this Dataframe, df :
 
 * <b>Player:</b> Name of the player 
 * <b>Span:</b> Duration of Test Career 
 * <b>Mat:</b> No. of matches played 
 * <b>Inns:</b> No. of innings played 
 * <b>Balls:</b> Total no. of balls bowled 
 * <b>Runs:</b> No. of runs conceded 
 * <b>Wkts:</b> No. of wickets taken altogether 
 * <b>BBI:</b> Best balling figure in an innings 
 * <b>BBM:</b> Best balling figure in a match (2 innings) 
 * <b>Ave:</b> Average meaning average no. of runs conceded per wicket 
 * <b>Econ:</b> Economy Rate, Econ= (Total runs conceded)/(Total over bowled)
 * <b>SR:</b> Strike Rate, SR means  the average no. of balls needed to bowl per wicket 
 * <b>5:</b> Shows the no. of 5-wicket wholes in an innings
 * <b>10:</b> Shows the number of times this bowler has taken ten wickets in a match 

<h4> No. of Rows and Columns </h4>

In [3]:
# No. of Rows and Columns in the dataframe, df
print("(row, column) -->", df.shape)

(row, column) --> (79, 14)


<h4>Remarks: </h4>

*   So, we can see there are 79 rows and 14 columns in the dataframe. Each row describes related information (Test Matches) about a Bowler. <br> <br>

<h4>Check for Missing Values</h4>

In [4]:
# Dataframe info, data types and checking the missing values
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79 entries, 0 to 78
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Player  79 non-null     object 
 1   Span    79 non-null     object 
 2   Mat     79 non-null     object 
 3   Inns    79 non-null     int64  
 4   Balls   79 non-null     int64  
 5   Runs    79 non-null     int64  
 6   Wkts    79 non-null     int64  
 7   BBI     79 non-null     object 
 8   BBM     79 non-null     object 
 9   Ave     79 non-null     float64
 10  Econ    79 non-null     float64
 11  SR      79 non-null     float64
 12  5       79 non-null     int64  
 13  10      79 non-null     int64  
dtypes: float64(3), int64(6), object(5)
memory usage: 8.8+ KB
None


<h4>Remarks:</h4>

* We see 79 observations (Players) and 14 variables (features).
* 3 floating point datatypes- Average, Economy and Strike Rate. <br>
* 'Inns', 'Balls', 'Runs', 'Wkts', '5', '10' : They contain integer values and <br>
* 'Player', 'Span', 'Mat', 'BBI', 'BBM' : Object (String/Mixed Data Type).<br>
* No missing values.<br>

<h4>Data Statistics</h4>

In [5]:
# Checking the data statistics
display(df.describe())

Inns         Balls          Runs        Wkts        Ave  \
count   79.000000     79.000000     79.000000   79.000000  79.000000   
mean   144.911392  18638.354430   8599.354430  317.215190  27.469747   
std     51.180222   7199.256972   3085.168807  121.924911   3.655658   
min     67.000000   8785.000000   4846.000000  200.000000  20.940000   
25%    110.000000  13583.000000   6456.500000  229.000000  24.500000   
50%    129.000000  16498.000000   7742.000000  266.000000  28.000000   
75%    169.000000  21742.500000   9756.000000  374.500000  29.870000   
max    304.000000  44039.000000  18355.000000  800.000000  34.790000   

            Econ         SR          5         10  
count  79.000000  79.000000  79.000000  79.000000  
mean    2.806835  59.193671  16.354430   2.797468  
std     0.351577   9.350132   9.642372   3.235935  
min     1.980000  41.200000   3.000000   0.000000  
25%     2.600000  53.300000   9.500000   1.000000  
50%     2.820000  57.400000  14.000000   2.000000  
75%     3.080000  63.950000  20.500000   3.500000  
max     3.460000  91.900000  67.000000  22.000000

<h4>Remarks: </h4>



 * Average no. of innings played by a bowler is around 144 with at least 67 innings.
 * Average Runs conceded by a bowler is 8599 where min is 4846.
 * It shows the records of bowlers who have taken at least 200 test wickets. <br> <br>
    

In [6]:
# shows the existing column headings of the dataframe

print(df.columns)

Index(['Player', 'Span', 'Mat', 'Inns', 'Balls', 'Runs', 'Wkts', 'BBI', 'BBM',
       'Ave', 'Econ', 'SR', '5', '10'],
      dtype='object')


<h4>Removing "BBI" and "BBM" from the dataset: </h4>

In [7]:
df.drop('BBM', axis=1, inplace=True)
df.drop('BBI', axis=1, inplace=True)

display(df.head())

Player       Span   Mat  Inns  Balls   Runs  Wkts    Ave  \
0  M Muralitharan (ICC/SL)  1992-2010   133   230  44039  18180   800  22.72   
1           SK Warne (AUS)  1992-2007   145   273  40705  17995   708  25.41   
2        JM Anderson (ENG)  2003-2021  164*   304  35079  16575   623  26.60   
3         A Kumble (INDIA)  1990-2008   132   236  40850  18355   619  29.65   
4         GD McGrath (AUS)  1993-2007   124   243  29248  12186   563  21.64   

   Econ    SR   5  10  
0  2.47  55.0  67  22  
1  2.65  57.4  37  10  
2  2.83  56.3  30   3  
3  2.69  65.9  35   8  
4  2.49  51.9  29   3

<h4>Renaming the column names</h4>

In [8]:
# Renaming the remaining column names

df.rename(columns={  'Span' : 'Career',
                    'Mat':'Match', 
                   'Inns':'Innings',
                    'Wkts': 'Wickets',
                    'Ave': 'Average',
                    'Econ': 'Economy',
                    'SR': 'Strike_rate',
                    '5': 'Five_wickets_wholes',
                    '10': 'Ten_wickets_in_a_match'}, inplace= True)

In [9]:
# displaying the first 5 rows of the dataframe which now have the changed column names

df.head(5)

Player     Career Match  Innings  Balls   Runs  Wickets  \
0  M Muralitharan (ICC/SL)  1992-2010   133      230  44039  18180      800   
1           SK Warne (AUS)  1992-2007   145      273  40705  17995      708   
2        JM Anderson (ENG)  2003-2021  164*      304  35079  16575      623   
3         A Kumble (INDIA)  1990-2008   132      236  40850  18355      619   
4         GD McGrath (AUS)  1993-2007   124      243  29248  12186      563   

   Average  Economy  Strike_rate  Five_wickets_wholes  Ten_wickets_in_a_match  
0    22.72     2.47         55.0                   67                      22  
1    25.41     2.65         57.4                   37                      10  
2    26.60     2.83         56.3                   30                       3  
3    29.65     2.69         65.9                   35                       8  
4    21.64     2.49         51.9                   29                       3

<h4>Counting the No. of players played for ICC</h4>

In [10]:
# Function created to be applied on dataset

def played_for_ICC(x):
    if 'ICC' in x:
        return "Yes"
    else:
        return "No"


# played_for_ICC is applied on the Dataframe df
df['played_for_ICC'] = df['Player'].apply(played_for_ICC)
display(df.head(10))

Player     Career Match  Innings  Balls   Runs  Wickets  \
0  M Muralitharan (ICC/SL)  1992-2010   133      230  44039  18180      800   
1           SK Warne (AUS)  1992-2007   145      273  40705  17995      708   
2        JM Anderson (ENG)  2003-2021  164*      304  35079  16575      623   
3         A Kumble (INDIA)  1990-2008   132      236  40850  18355      619   
4         GD McGrath (AUS)  1993-2007   124      243  29248  12186      563   
5          SCJ Broad (ENG)  2007-2021   149      274  29863  14590      524   
6            CA Walsh (WI)  1984-2001   132      242  30019  12688      519   
7            DW Steyn (SA)  2004-2019    93      171  18608  10077      439   
8      N Kapil Dev (INDIA)  1978-1994   131      227  27740  12867      434   
9        HMRKB Herath (SL)  1999-2018    93      170  25993  12157      433   

   Average  Economy  Strike_rate  Five_wickets_wholes  Ten_wickets_in_a_match  \
0    22.72     2.47         55.0                   67                      22   
1    25.41     2.65         57.4                   37                      10   
2    26.60     2.83         56.3                   30                       3   
3    29.65     2.69         65.9                   35                       8   
4    21.64     2.49         51.9                   29                       3   
5    27.84     2.93         56.9                   18                       3   
6    24.44     2.53         57.8                   22                       3   
7    22.95     3.24         42.3                   26                       5   
8    29.64     2.78         63.9                   23                       2   
9    28.07     2.80         60.0                   34                       9   

  played_for_ICC  
0            Yes  
1             No  
2             No  
3             No  
4             No  
5             No  
6             No  
7             No  
8             No  
9             No

In [11]:
# Counting the total number of players played for ICC 

total_count = df['played_for_ICC'].value_counts()
display(total_count)

No     74
Yes     5
Name: played_for_ICC, dtype: int64

<h4>Remarks:</h4>

* Here we can see only<b> 5</b> bowlers played for ICC and rest 74 didn't.

<h4>Getting Information About Different Countries from the Dataset</h4>

In [12]:
# Ignore all the warnings
import warnings
warnings.filterwarnings('ignore')

# Spliting the 'Player' column and removing "(" 
df_player = df['Player'].str.split("(", expand=True)

# Concatinating the extracted info with our main dataframe
df = pd.concat([df, df_player], axis=1)

# Dropping the 'Player' column as it appearred twice
df= df.drop('Player', axis=1)

# Renaming '0' and '1' columns as 'Player_Name' and 'Country' respectively
df = df.rename(columns={0: 'Player_Name',
                        1: 'Country'})

# Replacing ")" with ""
df['Country'] = df['Country'].str.replace(")", "")

# Rearranging the column names according to our wish
column_rearrange= ['Player_Name', 'Country','Career', 'Match', 'Innings', 'Balls', 'Runs', 'Wickets', 'Average',
                    'Economy', 'Strike_rate', 'Five_wickets_wholes',
                    'Ten_wickets_in_a_match', 'played_for_ICC']
df= df[column_rearrange]

display(df.head(10))

Player_Name Country     Career Match  Innings  Balls   Runs  Wickets  \
0  M Muralitharan   ICC/SL  1992-2010   133      230  44039  18180      800   
1        SK Warne      AUS  1992-2007   145      273  40705  17995      708   
2     JM Anderson      ENG  2003-2021  164*      304  35079  16575      623   
3        A Kumble    INDIA  1990-2008   132      236  40850  18355      619   
4      GD McGrath      AUS  1993-2007   124      243  29248  12186      563   
5       SCJ Broad      ENG  2007-2021   149      274  29863  14590      524   
6        CA Walsh       WI  1984-2001   132      242  30019  12688      519   
7        DW Steyn       SA  2004-2019    93      171  18608  10077      439   
8     N Kapil Dev    INDIA  1978-1994   131      227  27740  12867      434   
9    HMRKB Herath       SL  1999-2018    93      170  25993  12157      433   

   Average  Economy  Strike_rate  Five_wickets_wholes  Ten_wickets_in_a_match  \
0    22.72     2.47         55.0                   67                      22   
1    25.41     2.65         57.4                   37                      10   
2    26.60     2.83         56.3                   30                       3   
3    29.65     2.69         65.9                   35                       8   
4    21.64     2.49         51.9                   29                       3   
5    27.84     2.93         56.9                   18                       3   
6    24.44     2.53         57.8                   22                       3   
7    22.95     3.24         42.3                   26                       5   
8    29.64     2.78         63.9                   23                       2   
9    28.07     2.80         60.0                   34                       9   

  played_for_ICC  
0            Yes  
1             No  
2             No  
3             No  
4             No  
5             No  
6             No  
7             No  
8             No  
9             No

<h4>Counting the number of Players from different countries</h4>

In [13]:
# Removing 'ICC/' and '/ICC' from 'Country' column and displaying the count number
df['Country'] = df['Country'].str.replace("ICC/", "")
df['Country'] = df['Country'].str.replace("/ICC", "")
display(df['Country'].value_counts())

AUS      18
ENG      15
INDIA    10
WI        9
SA        8
NZ        7
PAK       7
SL        3
ZIM       1
BDESH     1
Name: Country, dtype: int64

<h4>Remarks:</h4>


* Here we can see, there are<b> 10 cricket playing nations</b> Australia, England, India, West Indies, South Africa, Pakistan, New Zealand, Sri Lanka, Zimbabwe and Bangladesh enlisted in the dataset. Now from this overview, questions may arise like:

* <b>Q:</b> How many <b>Australian</b> Bowlers are present in this dataset?<br><b>Answer:</b> 18
* <b>Q:</b> Is there any <b>Bangladeshi</b> player present in this dataset?<br><b>Answer:</b> Yes, there is only <b>1</b> player from Bangladesh in this dataset. <br> <br>

<h4>Analyzing the Test Career of The Bowlers</h4>

In [14]:
# splitting the 'Career' column and creating two new columns and changing their datatype from String to Integer
df['debut_year'] = df['Career'].str[0:4]
df['debut_year'] = df['debut_year'].astype('int')

df['retirement_year'] = df['Career'].str[5:]
df['retirement_year'] = df['retirement_year'].astype('int')

# removing the 'Career' column
df = df.drop('Career', axis=1)

# Career Duration = debut_year ~ retirement_year
df['career_duration'] = df['retirement_year'] - df['debut_year']

# Droping the column 'retirement_year'
df = df.drop('retirement_year', axis=1)

display(df.head())

Player_Name Country Match  Innings  Balls   Runs  Wickets  Average  \
0  M Muralitharan       SL   133      230  44039  18180      800    22.72   
1        SK Warne      AUS   145      273  40705  17995      708    25.41   
2     JM Anderson      ENG  164*      304  35079  16575      623    26.60   
3        A Kumble    INDIA   132      236  40850  18355      619    29.65   
4      GD McGrath      AUS   124      243  29248  12186      563    21.64   

   Economy  Strike_rate  Five_wickets_wholes  Ten_wickets_in_a_match  \
0     2.47         55.0                   67                      22   
1     2.65         57.4                   37                      10   
2     2.83         56.3                   30                       3   
3     2.69         65.9                   35                       8   
4     2.49         51.9                   29                       3   

  played_for_ICC  debut_year  career_duration  
0            Yes        1992               18  
1             No        1992               15  
2             No        2003               18  
3             No        1990               18  
4             No        1993               14

In [15]:
# Player with maximum duration of Test Career

sorted_maxDuration= df.sort_values(by=['career_duration'], ascending = False)
display(sorted_maxDuration.head())

Player_Name Country Match  Innings  Balls   Runs  Wickets  Average  \
21      Imran Khan      PAK    88      142  19458   8258      362    22.81   
55       GS Sobers       WI    93      159  21599   7999      235    34.03   
9     HMRKB Herath       SL    93      170  25993  12157      433    28.07   
0   M Muralitharan       SL   133      230  44039  18180      800    22.72   
2      JM Anderson      ENG  164*      304  35079  16575      623    26.60   

    Economy  Strike_rate  Five_wickets_wholes  Ten_wickets_in_a_match  \
21     2.54         53.7                   23                       6   
55     2.22         91.9                    6                       0   
9      2.80         60.0                   34                       9   
0      2.47         55.0                   67                      22   
2      2.83         56.3                   30                       3   

   played_for_ICC  debut_year  career_duration  
21             No        1971               21  
55             No        1954               20  
9              No        1999               19  
0             Yes        1992               18  
2              No        2003               18

<h4>Remarks:</h4>

* We can see from the above sorted representation that Pakistani Fast Bowler <b>Imran Khan had the longest Test career of 21 years</b> among all others available in the dataset follwed by <b>Sir Garry S Sobers (WI), HMRKB Herath (SL), Muttiah Muralitharan (SL)</b> and so on. <br><br>

In [16]:
# Player with minimum duration of Test Career

sorted_minDuration= df.sort_values(by=['career_duration'], ascending = True)
display(sorted_minDuration.head())

Player_Name Country Match  Innings  Balls  Runs  Wickets  Average  \
44      GP Swann      ENG    60      109  15349  7642      255    29.96   
71      K Rabada       SA    47       86   8785  4846      213    22.75   
54    Yasir Shah      PAK   46*       84  13607  7248      235    30.84   
61   SJ Harmison      ENG    63      115  13375  7192      226    31.82   
72  JR Hazlewood      AUS    55      103  11887  5438      212    25.65   

    Economy  Strike_rate  Five_wickets_wholes  Ten_wickets_in_a_match  \
44     2.98         60.1                   17                       3   
71     3.30         41.2                   10                       4   
54     3.19         57.9                   16                       3   
61     3.22         59.1                    8                       1   
72     2.74         56.0                    9                       0   

   played_for_ICC  debut_year  career_duration  
44             No        2008                5  
71             No        2015                6  
54             No        2014                7  
61            Yes        2002                7  
72             No        2014                7

<h4>Remarks:</h4>

* We can see from the above sorted representation that English Spinner<b> GP Swann had the shortest Test career of 5 years</b> among all others available in the dataset followed by <b>K Rabada (SA), Yasir Shah (PAK), SJ Harmison (ENG) </b> and so on. <br><br>

<h4>Player with the lowest Economy Rate</h4>

In [17]:
# Player with lowest economy rate

display(df.sort_values(by=['Economy']).head())

Player_Name Country Match  Innings  Balls  Runs  Wickets  Average  \
32      LR Gibbs       WI    79      148  27115  8989      309    29.09   
47      R Benaud      AUS    63      116  19108  6704      248    27.03   
35  DL Underwood      ENG    86      151  21862  7674      297    25.83   
39       BS Bedi    INDIA    67      118  21364  7637      266    28.71   
68   CV Grimmett      AUS    37       67  14513  5231      216    24.21   

    Economy  Strike_rate  Five_wickets_wholes  Ten_wickets_in_a_match  \
32     1.98         87.7                   18                       2   
47     2.10         77.0                   16                       1   
35     2.10         73.6                   17                       6   
39     2.14         80.3                   14                       1   
68     2.16         67.1                   21                       7   

   played_for_ICC  debut_year  career_duration  
32             No        1958               18  
47             No        1952               12  
35             No        1966               16  
39             No        1966               13  
68             No        1925               11

<h4>Remarks:</h4>

* <b>LR Gibbs (WI)</b> has the lowest economy rate. He conceeded just 8989 runs in around 4519.17 overs with the lowest economy rate of 1.98. <br><br>

<h4>Player with the lowest Strike Rate</h4>

In [18]:
# Player with lowest strike rate

display(df.sort_values(by=['Strike_rate']).head())

Player_Name Country Match  Innings  Balls   Runs  Wickets  Average  \
71      K Rabada       SA    47       86   8785   4846      213    22.75   
7       DW Steyn       SA    93      171  18608  10077      439    22.95   
20  Waqar Younis      PAK    87      154  16224   8788      373    23.56   
19   MD Marshall       WI    81      151  17584   7876      376    20.94   
25     AA Donald       SA    72      129  15519   7344      330    22.25   

    Economy  Strike_rate  Five_wickets_wholes  Ten_wickets_in_a_match  \
71     3.30         41.2                   10                       4   
7      3.24         42.3                   26                       5   
20     3.25         43.4                   22                       5   
19     2.68         46.7                   22                       4   
25     2.83         47.0                   20                       3   

   played_for_ICC  debut_year  career_duration  
71             No        2015                6  
7              No        2004               15  
20             No        1989               14  
19             No        1978               13  
25             No        1992               10

<h4>Remarks:</h4>

* <b>K Rabada</b> has the lowest strike rate. He preyed after every 41 balls (approximately). <br><br>

<h4>Player with the lowest Bowling Average</h4>

In [19]:
# Player with lowest bowling average

display(df.sort_values(by=['Average']).head())

Player_Name Country Match  Innings  Balls   Runs  Wickets  Average  \
19  MD Marshall       WI    81      151  17584   7876      376    20.94   
41     J Garner       WI    58      111  13169   5433      259    20.97   
15  CEL Ambrose       WI    98      179  22103   8501      405    20.99   
33   FS Trueman      ENG    67      127  15178   6625      307    21.57   
4    GD McGrath      AUS   124      243  29248  12186      563    21.64   

    Economy  Strike_rate  Five_wickets_wholes  Ten_wickets_in_a_match  \
19     2.68         46.7                   22                       4   
41     2.47         50.8                    7                       0   
15     2.30         54.5                   22                       3   
33     2.61         49.4                   17                       3   
4      2.49         51.9                   29                       3   

   played_for_ICC  debut_year  career_duration  
19             No        1978               13  
41             No        1977               10  
15             No        1988               12  
33             No        1952               13  
4              No        1993               14

<h4>Remarks:</h4>

* <b>MD marshall</b> has the lowest bowling average. He conceeded approximately 21 runs on average to take a test wicket in his career. <br><br>